openhcl/virt_mshv_vtl/src/processor/snp/mod.rs

```diff
@@ -929,6 +972,193 @@ impl UhProcessor<'_, SnpBacked> {
+            cpl: vmsa.cpl(),
+        }) {
+            if self.cvm_protect_secure_register_write(
+                entered_from_vtl,
+                HvX64RegisterName::Xfem,
+                value,
+            ) {
+                // Once the intercept message has been posted, no further
+                // processing is required. Do not advance the instruction
+                // pointer here, since the instruction pointer must continue to
+                // point to the instruction that generated the intercept.
+            } else {
+                let mut vmsa = self.runner.vmsa_mut(entered_from_vtl);
+                vmsa.set_xcr0(value);
+                advance_to_next_instruction(&mut vmsa);
+            }
+        } else {
+            let mut vmsa = self.runner.vmsa_mut(entered_from_vtl);
+            vmsa.set_event_inject(
+                SevEventInjectInfo::new()
+                    .with_interruption_type(x86defs::snp::SEV_INTR_TYPE_EXCEPT)
+                    .with_vector(x86defs::Exception::GENERAL_PROTECTION_FAULT.0)
+                    .with_deliver_error_code(true)
+                    .with_valid(true),
+            );
+        }
+    }
+
+    fn handle_crx_intercept(&mut self, entered_from_vtl: GuestVtl, reg: HvX64RegisterName) {
+        let vmsa = self.runner.vmsa(entered_from_vtl);
+        let mov_crx_drx = x86defs::snp::MovCrxDrxInfo::from(vmsa.exit_info1());
+        let reg_value = {
+            let gpr_name =
+                HvX64RegisterName(HvX64RegisterName::Rax.0 + mov_crx_drx.gpr_number() as u32);
+
+            match gpr_name {
+                HvX64RegisterName::Rax => vmsa.rax(),
+                HvX64RegisterName::Rbx => vmsa.rbx(),
+                HvX64RegisterName::Rcx => vmsa.rcx(),
+                HvX64RegisterName::Rdx => vmsa.rdx(),
+                HvX64RegisterName::Rsp => vmsa.rsp(),
+                HvX64RegisterName::Rbp => vmsa.rbp(),
+                HvX64RegisterName::Rsi => vmsa.rsi(),
+                HvX64RegisterName::Rdi => vmsa.rdi(),
+                HvX64RegisterName::R8 => vmsa.r8(),
+                HvX64RegisterName::R9 => vmsa.r9(),
+                HvX64RegisterName::R10 => vmsa.r10(),
+                HvX64RegisterName::R11 => vmsa.r11(),
+                HvX64RegisterName::R12 => vmsa.r12(),
+                HvX64RegisterName::R13 => vmsa.r13(),
+                HvX64RegisterName::R14 => vmsa.r14(),
+                HvX64RegisterName::R15 => vmsa.r15(),
+                _ => unreachable!("unexpected register"),
+            }
+        };
+
+        // Special case: LMSW/CLTS/SMSW intercepts do not provide decode assist
+        // information. No support to emulate these instructions yet, but the
+        // access by the guest might be allowed by the higher VTL and therefore
+        // crashing is not necessarily the correct behavior.
+        //
+        // TODO SNP: consider emulating the instruction.
+        if !mov_crx_drx.mov_crx() {
```

Comment:
I'll note that we have some basic mov crx emulation code in TDX already. So we could probably generalize that and use it here.